tests/unit/s2n_client_cert_verify_test.c

```diff
@@ -42,6 +46,86 @@ static int test_sign(const struct s2n_pkey *priv_key, s2n_signature_algorithm si
     return S2N_SUCCESS;
 }
 
+struct task_params {
+    struct s2n_connection *conn;
+    struct s2n_async_pkey_op *op;
+};
+
+void *pkey_task(void *param)
+{
+    struct task_params *params = (struct task_params*)param;
+
+    struct s2n_cert_chain_and_key *chain_and_key = s2n_connection_get_selected_cert(params->conn);
+    EXPECT_NOT_NULL(chain_and_key);
+
+    s2n_cert_private_key *pkey = s2n_cert_chain_and_key_get_private_key(chain_and_key);
+    EXPECT_NOT_NULL(pkey);
+
+    EXPECT_SUCCESS(s2n_async_pkey_op_perform(params->op, pkey));
+
+    free(params);
+    pthread_exit(NULL);
+}
+
+int async_pkey_perform_op(struct s2n_connection *conn, struct s2n_async_pkey_op *op)
+{
+    EXPECT_NOT_NULL(conn);
+    EXPECT_NOT_NULL(op);
+
+    pkey_op = op;
+
+    struct task_params *params = malloc(sizeof(struct task_params));
+    EXPECT_NOT_NULL(params);
+    
+    params->conn = conn; 
+    params->op = op; 
+
+    POSIX_GUARD(pthread_create(&worker, NULL, &pkey_task, params));
+
+    return S2N_SUCCESS;
+}
+
+static int try_handshake(struct s2n_connection *server_conn, struct s2n_connection *client_conn)
+{
+    s2n_blocked_status server_blocked = {0};
+    s2n_blocked_status client_blocked = {0};
+
+    int server_rc;
+    int client_rc;
+
+    do {
+        client_rc = s2n_negotiate(client_conn, &client_blocked);
+        if (!(client_rc == 0 || (client_blocked && s2n_error_get_type(s2n_errno) == S2N_ERR_T_BLOCKED))) {
+            return S2N_FAILURE;
+        }
+
+        if(client_blocked == S2N_BLOCKED_ON_APPLICATION_INPUT && pkey_op != NULL) {
+            if(s2n_async_pkey_op_apply(pkey_op, client_conn) == S2N_SUCCESS) {
+                POSIX_GUARD(pthread_join(worker, NULL));
+                EXPECT_SUCCESS(s2n_async_pkey_op_free(pkey_op));
+                pkey_op = NULL;
+            }
+        }
+
+        server_rc = s2n_negotiate(server_conn, &server_blocked);
+        if (!(server_rc == 0 || (server_blocked && s2n_error_get_type(s2n_errno) == S2N_ERR_T_BLOCKED))) {
+            return S2N_FAILURE;
+        }
+
+        if(server_blocked == S2N_BLOCKED_ON_APPLICATION_INPUT && pkey_op != NULL) {
+            if(s2n_async_pkey_op_apply(pkey_op, server_conn) == S2N_SUCCESS) {
+                POSIX_GUARD(pthread_join(worker, NULL));
```

Comment:
pthread_join waits for the thread to complete. Would it make more sense to call join BEFORE calling apply on the op? That way you won't attempt to apply until the thread has completed.

Reply:
Yes will make this change